src/components/Form/FormProvider.tsx

```diff
@@ -0,0 +1,359 @@
+                            if (
+                                relatedTargetId === CONST.OVERLAY.BOTTOM_BUTTON_NATIVE_ID ||
+                                relatedTargetId === CONST.OVERLAY.TOP_BUTTON_NATIVE_ID ||
+                                relatedTargetId === CONST.BACK_BUTTON_NATIVE_ID
+                            ) {
+                                return;
+                            }
+                            setTouchedInput(inputID);
+                            if (shouldValidateOnBlur) {
+                                onValidate(inputValues, !hasServerError);
+                            }
+                        }, VALIDATE_DELAY);
+                    }
+                    inputProps.onBlur?.(event);
+                },
+                onInputChange: (value: unknown, key?: string) => {
+                    const inputKey = key ?? inputID;
+                    setInputValues((prevState) => {
+                        const newState = {
+                            ...prevState,
+                            [inputKey]: value,
+                        };
+
+                        if (shouldValidateOnChange) {
+                            onValidate(newState);
+                        }
+                        return newState;
+                    });
+
+                    if (inputProps.shouldSaveDraft && !(formID as string).includes('Draft')) {
+                        FormActions.setDraftValues(formID as OnyxFormKeyWithoutDraft, {[inputKey]: value});
+                    }
+                    inputProps.onValueChange?.(value, inputKey);
+                },
+            };
+        },
+        [draftValues, formID, errors, formState, hasServerError, inputValues, onValidate, setTouchedInput, shouldValidateOnBlur, shouldValidateOnChange],
+    );
+    const value = useMemo(() => ({registerInput}), [registerInput]);
+
+    return (
+        <FormContext.Provider value={value}>
+            {/* eslint-disable react/jsx-props-no-spreading */}
+            <FormWrapper
+                {...rest}
+                formID={formID}
+                onSubmit={submit}
+                inputRefs={inputRefs}
+                errors={errors}
+                enabledWhenOffline={enabledWhenOffline}
+            >
+                {typeof children === 'function' ? children({inputValues}) : children}
+            </FormWrapper>
+        </FormContext.Provider>
+    );
+}
+
+FormProvider.displayName = 'Form';
+
+export default withOnyx<FormProviderProps, FormProviderOnyxProps>({
+    network: {
+        key: ONYXKEYS.NETWORK,
+    },
+    formState: {
+        // @ts-expect-error TODO: fix this
+        key: ({formID}) => formID,
+    },
+    draftValues: {
+        // @ts-expect-error TODO: fix this
+        key: (props) => `${props.formID}Draft` as const,
+    },
+})(forwardRef(FormProvider)) as <TFormID extends OnyxFormKey>(props: Omit<FormProviderProps<TFormID>, keyof FormProviderOnyxProps>) => ReactNode;
```

Comment:
```suggestion
export default withOnyx<FormProviderProps, FormProviderOnyxProps>({
    network: {
        key: ONYXKEYS.NETWORK,
    },
    formState: {
        // Comment explaining..
        // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-explicit-any
        key: ({formID}) => formID as any,
    },
    draftValues: {
        // Comment explaining..
        // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-explicit-any
        key: (props) => `${props.formID}Draft` as any,
    },
})(forwardRef(FormProvider)) as <TFormID extends OnyxFormKey>(props: Omit<FormProviderProps<TFormID>, keyof FormProviderOnyxProps>) => ReactNode;
```

We can do the same for this one since it's a generic component, but we'll have to fix those props to use `Form`.

```ts
// src/components/Form/FormProvider.tsx
type FormProviderOnyxProps = {
    /** Contains the form state that must be accessed outside the component */
    formState: OnyxEntry<Form>;

    /** Contains draft values for each input in the form */
    draftValues: OnyxEntry<Form>;

    /** Information about the network */
    network: OnyxEntry<Network>;
};
```

And we can change `Form` type to allow unknown keys.

```ts
// src/types/onyx/Form.ts
type Form = {
    [key: string]: unknown;

    /** Controls the loading state of the form */
    isLoading?: boolean;

    /** Server side errors keyed by microtime */
    errors?: OnyxCommon.Errors | null;

    /** Field-specific server side errors keyed by microtime */
    errorFields?: OnyxCommon.ErrorFields | null;
};
```

WDYT @blazejkustra @kowczarz ?